File: Code/R/dataSetGetter.R
# Function for getting data sets within a range of dates
# This function returns a data frame consisting of the data sets within the range [startDate, endDate]
# The start and end dates are inclusive
# The directory parameter is the path to the directory where the data sets are stored. This parameter should end with "/"
# The parameter onlyActualTravelTimes is optional. However, if the argument is provided and it is TRUE, then only the column representing
# the actual travel times is returned
getDataSet <- function(startDate, endDate, directory, model, onlyActualTravelTimes=FALSE){
  # Retrieve all files in directory
  fileNames = list.files(directory)
  # Filter out predictions from other models
  print(model)
  fileNames <- fileNames[grep(model, fileNames)]
  # Assume that the eight first characters are the date for the respective data set
  fileDates = substr(fileNames, 1, 8)
  # Convert dates to date objects
  fileDates = as.Date(fileDates, "%Y%m%d")
  # Combine fileNames and fileDates into one data frame
  files = data.frame(fileNames, fileDates)
  # Convert input dates to date objects
  startDate = as.Date(startDate, "%Y%m%d")
  endDate = as.Date(endDate, "%Y%m%d")
  # Select rows having dates within the given range
  files = files[(files$fileDates >= startDate) & (files$fileDates <= endDate),]
  # Read data sets within the given range, and store them in one data frame
  firstFileName = files$fileNames[1]
  combinedDataSet = data.frame(read.csv(paste(directory, firstFileName, sep=""), sep=";", stringsAsFactor = FALSE))
  for(fileName in files$fileNames[-1]){
    dataSet = data.frame(read.csv(paste(directory, fileName, sep=""), sep=";", stringsAsFactor = FALSE))
    combinedDataSet = data.frame(rbind(combinedDataSet, dataSet))
  }
  if(!missing(onlyActualTravelTimes) & onlyActualTravelTimes){
    combinedDataSet = data.frame(combinedDataSet$actualTravelTime)
  }
  return(combinedDataSet)
}

# Returns a data set for which each column represents the predictions for the different baselines provided in the argument "baselines"
# This function assumes that directory is the path to the dataset folder, and that there exists a sub-folder in that directory
# for each baseline which contains predictions made from that baseline for the different dates
getDataSetForBaselines <- function(startDate, endDate, directory, baselines){
  firstBaseLine = baselines[1]
  combinedDataSet = data.frame(getDataSet(startDate, endDate, paste(directory, firstBaseLine, "/", sep="")))
  for(baseline in baselines[-1]){
    dataSet = data.frame(getDataSet(startDate, endDate, paste(directory, baseline, "/", sep="")))
    combinedDataSet = data.frame(cbind(combinedDataSet, dataSet))
  }
  colnames(combinedDataSet) = baselines
  return(combinedDataSet)
}

# Function for normalizing a vector
normalize <- function(x, min, max){
  return((x-min)/(max-min))
}

# Function for denormalizing a vector
deNormalize <- function(x, min, max){
  return((x*(max-min))+min)
}

storeDataSet <- function(dataSet, directory) {
  #create list of dates
  firstDate <- as.Date(dataSet[1, "dateAndTime"])
  lastDate <- as.Date(dataSet[nrow(dataSet), "dateAndTime"])
  listOfDates <- seq(firstDate, lastDate, by="days")
  
  #create data frame from dataSet for each day in list of dates and write to csv file
  for (i in 1:length(listOfDates)) {
    date = listOfDates[i]
    write.table(dataSet[dataSet$dateAndTime == date, colnames(dataSet)], file = paste(directory, gsub("-", "", as.character(date)), "_filteredDataSet.csv", sep = ""), sep = ";", row.names=FALSE)
  }
}

File: Code/Python/dataSetGetter.py
import numpy as np
from datetime import *
from os import listdir
from os.path import isfile, join

# Function for converting string to date time object
dateTimeConverter = lambda x: datetime.strptime(x, '%Y-%m-%d %H:%M:%S')

def getDataSet(startDate, endDate, directory, model, onlyActualTravelTimes=False):
	# Convert date strings to datetime objects
	startDate = datetime.strptime(startDate, '%Y%m%d')
	endDate = datetime.strptime(endDate, '%Y%m%d')

	# Retrieve file names in directory and only select the ones containing model
	filenames = [filename for filename in listdir(directory) if isfile(join(directory,filename)) and model in filename]

	# Retrieve dates for the files in the directory
	dates = [datetime.strptime(filename[0:8], '%Y%m%d') for filename in filenames]

	# Filenames with dates within the range of the provided dates
	dataSetFileNames = [filename for (filename, date) in zip(filenames, dates) if (date>=startDate) and (date<=endDate)]

	# Initialize array for containing the data set
	dataSet = np.zeros((1,), dtype=[('dateAndTime', datetime), ('fiveMinuteMean', np.float), ('trafficVolume', np.int), ('actualTravelTime', np.float)])

	# Store all data sets for the dates in the provided range in one array
	for dataSetFileName in dataSetFileNames:
		# Read data set from file
		dataSetForOneDate = np.genfromtxt(join(directory, dataSetFileName), delimiter=';', missing_values=0, skip_header=0, dtype=(datetime, np.float, np.int, np.float), usecols=(0, 1, 2, 3), names=True, converters={0: dateTimeConverter})
		
		# Add data set to matrix of data
		dataSet = np.concatenate([dataSet, dataSetForOneDate])

	# Remove first row because it is just zeros
	dataSet = np.delete(dataSet, 0, 0)

	# Return only actual travel times if flag is set, otherwise return whole data set
	return dataSet['actualTravelTime'] if onlyActualTravelTimes else dataSet

def saveDataSet(directory, filename, dataSet, format, header):
	np.savetxt(join(directory, filename), dataSet, fmt=format, header=header, comments='')	
# np.savetxt(join(directory, '20150129_dataset_test.csv'), dataSet, fmt=('%s;%f;%i;%f'), header='dateAndTime;fiveMinuteMean;trafficVolume;actualTravelTime', comments='')